### PR TITLE
DAOS-13881 control: Clean associated handles on agent start

### PR DESCRIPTION
**Pull request by tanabarr**
In the agent there is a process monitoring feature (procmon.go) that
tracks application connections (pool handles). It calls PoolEvict for
the handles that an application has created when the application
process no longer exists. It has been perceived that there is a gap
where dangling handles could be left in the case of a host crash where
the agent exit handlers are not able to execute and revoke relevant
application-created pool handles as expected. To address this gap this
change adds a routine to call SystemCleanup(<machine_name>) on agent
start-up which will revoke any pool handles associated with the host
the agent is running on.

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by tanabarr** [comment]
See https://daosio.atlassian.net/browse/DAOS-13881?focusedCommentId=125642 for details of PR manual verification.

File: src/control/cmd/daos_agent/procmon.go [hunk 1]
```diff
@@ -342,5 +345,53 @@ func (p *procMon) handleRequests(ctx context.Context) {
...
+	}
+	err = resp.Errors()
+	if err != nil {
+		p.log.Errorf("%s: system cleanup ran with error: %s", msg, err)
+		return
+	}
+
+	msg = fmt.Sprintf("Running system cleanup on this %s", msg)
+	if len(resp.Results) == 0 {
+		p.log.Infof("%s: no pool handles revoked", msg)
```
Comment (mjmac) [inline src/control/cmd/daos_agent/procmon.go:386]: Probably better as debug, no? I don't think this is relevant to normal operations.

File: utils/config/daos_agent.yml [hunk 2]
```diff
@@ -65,7 +65,15 @@
...
 ## in interruption of DAOS I/O for any local DAOS client processes that have
 ## an open pool handle.
-# disable_auto_evict: true
+## default: false
+#disable_auto_evict: true
+
+## Enable automatic eviction of open pool handles on agent startup. If enabled (disabled by
+## default), the agent will evict all open pool handles associated with the machine/host name
+## on agent start. This aims to garbage collect any handles not evicted after an agent or host
+## crash.
```
Comment (mjmac) [inline utils/config/daos_agent.yml:74]: Minor wording suggestion to remove some redundancy and irrelevant (to the user) details:
```suggestion
## If enabled, the agent will evict any open pool handles associated with this machine
## on agent startup. This allows the servers to reclaim resources that may not have been
## properly cleaned up in the event of an agent or machine crash.
```

File: src/control/cmd/daos_agent/procmon.go [hunk 3]
```diff
@@ -342,5 +345,53 @@ func (p *procMon) handleRequests(ctx context.Context) {
 // startMonitoring is the main driver which starts the process monitor. The
 // passed in context is used to terminate all monitoring in the event of shutdown.
 func (p *procMon) startMonitoring(ctx context.Context) {
+	if p.cleanOnStart {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/cmd/daos_agent/procmon.go:348]: done

File: src/control/cmd/daos_agent/procmon.go [hunk 4]
```diff
@@ -342,5 +345,53 @@ func (p *procMon) handleRequests(ctx context.Context) {
...
+	if err != nil {
+		p.log.Errorf("hostname lookup: %s, cannot cleanup handles", err)
+		return
+	}
+	// Strip the domain off of the Hostname
+	machineName := strings.Split(hostName, ".")[0]
+	if machineName == "" {
+		p.log.Errorf("empty machine name is invalid, cannot cleanup handles", err)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/cmd/daos_agent/procmon.go:367]: done

File: src/control/cmd/daos_agent/procmon.go (hunk 1 above)
Comment (tanabarr) [reply src/control/cmd/daos_agent/procmon.go:386]: done

File: utils/config/daos_agent.yml (hunk 2 above)
Comment (tanabarr) [reply utils/config/daos_agent.yml:74]: done

File: utils/config/daos_agent.yml [hunk 5]
```diff
@@ -65,7 +65,14 @@
...
 ## an open pool handle.
-# disable_auto_evict: true
+## default: false
+#disable_auto_evict: true
+
+## If enable, the agent will evict any open pool handles associated with this machine on agent
+## startup. This allows the servers to reclaim resources that may not have been properly cleaned
+## up in the event of an agent or machine crash.
+## default: false
+#clean_handles_on_start: true
```
[the remark this answers is not in the document]
Comment (liw) [reply utils/config/daos_agent.yml:75]: Of course, my assumption is that the new option **is** inherent similar to the existing one. If that's false, my comment should be ignored.

[the remark this answers is not in the document]
Comment (tanabarr) [reply utils/config/daos_agent.yml:75]: done

[the remark this answers is not in the document]
Comment (liw) [reply utils/config/daos_agent.yml:75]: The comments sound like if both should include "auto". The desired default behavior of not doing so upon start can also be achieved via a default value of true for `disable_auto_evict_on_start`, I think.

[the remark this answers is not in the document]
Comment (tanabarr) [reply utils/config/daos_agent.yml:75]: As the behaviour is slightly different I think `enable_evict_on_start` is appropriate. It is a subtle difference but "auto_evict" to me refers to the procmon.go tracking of application handles and eviction when the application process dies whereas "evict_on_start" is a one off evict operation on start-up.

**Review by liw — APPROVED** [review]

File: utils/config/daos_agent.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (liw) [reply utils/config/daos_agent.yml]: OK, sounds good to me. Thanks, Tom.

File: src/control/security/auth/auth_sys.go [hunk 6]
```diff
@@ -146,6 +146,16 @@ func sysNameToPrincipalName(name string) string {
 	return name + "@"
 }
 
+func GetMachineName() (string, error) {
```
Comment (mjmac) [inline src/control/security/auth/auth_sys.go:149]: minor: This is a public function, so it should have a docstring.

**Review by mjmac — APPROVED** [review]
Updates LGTM.

File: src/control/cmd/daos_agent/procmon.go [hunk 7]
```diff
@@ -341,6 +342,52 @@ func (p *procMon) handleRequests(ctx context.Context) {
...
+	msgRvkd := ""
+	for i, r := range resp.Results {
+		msgRvkd = fmt.Sprintf("%s%d handles revoked for pool %q", msgRvkd, r.Count, r.PoolID)
+		if i != len(resp.Results)-1 {
+			msgRvkd += ", "
+		}
+	}
+	p.log.Infof("%s: %s", msg, msgRvkd)
```
Comment (kjacque) [inline src/control/cmd/daos_agent/procmon.go:392]: Just a suggestion, but maybe this log message could be made a bit less verbose, something like "poolID (count)" for each pool. The overall message could look something like:
`Pool handles revoked: pool1 (3), pool5 (10), pool6 (1)`

**Review by kjacque — DISMISSED** [review]
Nothing I'd block on here, good work!

**Review by knard-intel — DISMISSED** [review]
LGTM.

File: src/control/cmd/daos_agent/procmon.go (hunk 7 above)
Comment (knard-intel) [reply src/control/cmd/daos_agent/procmon.go:392]: This indeed, could be helpful for support tasks.

Comment (tanabarr) [reply src/control/cmd/daos_agent/procmon.go:392]: Yes this is much better, thanks. Done

File: src/control/security/auth/auth_sys.go (hunk 6 above)
Comment (mjmac) [reply src/control/security/auth/auth_sys.go:150]: nit: The point I was trying to make with my other comment was that the format of a machine name is an implementation detail which could change, hence the advice to refactor into a common implementation that will ensure compatibility across code that needs to get at a machine name that will be or was stored in a pool handle.

It's perhaps a subtle distinction, but for a helper like this, IMO the docstring shouldn't describe the implementation, but the purpose/intended use of the function. In this case, the intended use is to generate a machine name suitable for storage in a pool handle. If you wanted to go overboard (I don't recommend this), then you would probably add a utility function called `GetShortHostname()` which does as it says on the tin, and then `GetMachineName()` would call that function. That way if the requirements for what gets stored in the pool handle change, you only need to change the implementation of `GetMachineName()` and callers don't know/don't care about the details. A suitably descriptive-but-generalized docstring wouldn't need to be updated, either.

Not asking for a change here, just some food for thought.

**Review by mjmac — APPROVED** [review]

**Review by knard-intel — APPROVED** [review]

**Comment by tanabarr** [comment]
Pylint failures unrelated to PR

**Pull request merged**